File: javascriptv3/example_code/cloudwatch-logs/scenarios/large-query/scouts/rust/cloud-watch-query.rs
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
use aws_config::SdkConfig;
use aws_sdk_cloudwatchlogs::{Client, Error, GetQueryResultsRequest, StartQueryRequest};
use std::time::{Duration, SystemTime};

struct QueryConfig {
    log_group_names: Vec<String>,
    date_range: (SystemTime, SystemTime),
    limit: u32,
}

struct QueryResults {
    results: Vec<Vec<String>>,
    seconds_elapsed: u64,
}

impl QueryResults {
    fn new(results: Vec<Vec<String>>, elapsed: Duration) -> Self {
        QueryResults {
            results,
            seconds_elapsed: elapsed.as_secs(),
        }
    }
}

struct Query {
    client: Client,
    config: QueryConfig,
}

impl Query {
    fn new(config: SdkConfig, query_config: QueryConfig) -> Self {
        let client = Client::new(&config);
        Query { client, config } 
    }

    async fn run(&self) -> QueryResults {
        let start = SystemTime::now();
        let results = self.query_logs(self.config.date_range).await;
        let end = SystemTime::now();
        let elapsed = end.duration_since(start).unwrap();
        
        QueryResults::new(results, elapsed)
    }

    async fn query_logs(&self, date_range: (SystemTime, SystemTime)) -> Vec<Vec<String>> {
        // Implementation of recursive query
        todo!() 
    }

    async fn start_query(&self, date_range: (SystemTime, SystemTime)) -> Result<String, Error> {
        let start_ms = date_range.0.duration_since(SystemTime::UNIX_EPOCH)?.as_millis() as i64;
        let end_ms = date_range.1.duration_since(SystemTime::UNIX_EPOCH)?.as_millis() as i64;

        let request = StartQueryRequest::builder()
            .log_group_names(self.config.log_group_names.clone())
            .start_time(start_ms)
            .end_time(end_ms)
            .limit(self.config.limit)
            .query_string("fields @timestamp, @message | sort @timestamp asc".into())
            .build();

        self.client.start_query(request).await.map(|resp| resp.query_id)
    }

    async fn get_query_results(&self, query_id: &str) -> Result<Option<Vec<Vec<String>>>, Error> {
        let request = GetQueryResultsRequest::builder()
            .query_id(query_id.to_string())
            .build();

        let response = self.client.get_query_results(request).await?;

        Ok(response.results)
    }
}

File: javascriptv3/example_code/cloudwatch-logs/scenarios/large-query/scouts/python/cloud-watch-query.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0
import boto3
from botocore.config import Config
from datetime import datetime, timedelta
import time


class DateOutOfBoundsError(Exception):
    pass


class CloudWatchQuery:
    def __init__(self, log_group_names, date_range, limit=10000):
        self.logs_client = boto3.client(
            "logs", config=Config(retries={"max_attempts": 10})
        )
        self.log_group_names = log_group_names
        self.date_range = date_range
        self.limit = limit
        self.results = []
        self.seconds_elapsed = None

    def run(self):
        start = datetime.now()
        self.results = self._large_query(self.date_range)
        end = datetime.now()
        self.seconds_elapsed = (end - start).total_seconds()
        return self.results

    def _large_query(self, date_range):
        logs = self._query(date_range, self.limit)

        print(
            f"Query date range: {date_range[0].isoformat()} to {date_range[1].isoformat()}. Found {len(logs)} logs."
        )

        if len(logs) < self.limit:
            return logs

        last_log_date = self._get_last_log_date(logs)
        offset_last_log_date = last_log_date + timedelta(milliseconds=1)
        sub_date_range = [offset_last_log_date, date_range[1]]
        r1, r2 = self._split_date_range(sub_date_range)

        results = []
        results.extend(self._large_query(r1))
        results.extend(self._large_query(r2))

        return [logs] + results

    def _get_last_log_date(self, logs):
        timestamps = [log["@timestamp"] for log in logs if "@timestamp" in log]
        timestamps.sort()
        return datetime.fromisoformat(timestamps[-1] + "Z")

    def _split_date_range(self, date_range):
        midpoint = date_range[0] + (date_range[1] - date_range[0]) / 2
        return [date_range[0], midpoint], [midpoint, date_range[1]]

    def _query(self, date_range, max_logs):
        try:
            query_id = self._start_query(date_range, max_logs)
            results = self._wait_until_query_done(query_id)
            return results["results"] if "results" in results else []
        except DateOutOfBoundsError:
            return []

    def _start_query(self, date_range, max_logs):
        try:
            response = self.logs_client.start_query(
                logGroupName=self.log_group_names,
                startTime=int(date_range[0].timestamp() * 1000),
                endTime=int(date_range[1].timestamp() * 1000),
                queryString="fields @timestamp, @message | sort @timestamp asc",
                limit=max_logs,
            )
            return response["queryId"]
        except self.logs_client.exceptions.ResourceNotFoundException as e:
            raise DateOutOfBoundsError(e)

    def _get_query_results(self, query_id):
        return self.logs_client.get_query_results(queryId=query_id)

    def _wait_until_query_done(self, query_id):
        while True:
            time.sleep(1)
            results = self._get_query_results(query_id)
            if results["status"] in [
                "Complete",
                "Failed",
                "Cancelled",
                "Timeout",
                "Unknown",
            ]:
                return results
